File: app/src/main/java/com/kooritea/fcmfix/xposed/AutoStartFix.java
package com.kooritea.fcmfix.xposed;

import android.content.Intent;

import com.kooritea.fcmfix.util.XposedUtils;

import java.lang.reflect.Method;

import com.kooritea.fcmfix.libxposed.XC_MethodHook;
import com.kooritea.fcmfix.libxposed.XposedBridge;
import com.kooritea.fcmfix.libxposed.XposedHelpers;

/**
 * AutoStartFix - 自启动修复模块
 * 
 * 功能说明：
 * 解决各 ROM 系统对应用自启动的限制，允许 FCM/GCM 推送消息能够唤醒已停止的应用。
 * 主要针对不同 Android 版本和厂商定制系统（MIUI、HyperOS、ColorOS、OxygenOS）的自启动管理策略进行 Hook。
 * 
 * 工作原理：
 * 1. Hook 系统广播发送前的检查方法，绕过自启动限制
 * 2. 当检测到 FCM 相关 Intent 时，强制允许应用接收广播
 * 3. 支持多个系统版本的不同类和方法名
 * 
 * 配合 fcmfix 可实现无代理直接接收 FCM 消息的关键组件之一
 */
public class AutoStartFix extends XposedModule {
    // FCM 接收广播的标准 Action
    private final String FCM_RECEIVE = ".android.c2dm.intent.RECEIVE";

    public AutoStartFix(ClassLoader classLoader){
        super(classLoader);
        try{
            // Hook 自启动检查方法
            this.startHook();
            // Hook 移除电源策略限制
            this.startHookRemovePowerPolicy();
        }catch (Throwable e) {
            printLog("hook error AutoStartFix:" + e.getMessage());
        }
    }

    /**
     * 开始 Hook 各个系统的自启动检查方法
     * 覆盖了从 MIUI 12 到 HyperOS 以及 OOS/ColorOS 15 的多种实现
     */
    protected void startHook(){
        try{
            // MIUI 12: Hook BroadcastQueueInjector.checkApplicationAutoStart
            Class<?> BroadcastQueueInjector = XposedHelpers.findClass("com.android.server.am.BroadcastQueueInjector",classLoader);
            XposedUtils.findAndHookMethodAnyParam(BroadcastQueueInjector,"checkApplicationAutoStart",new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    Intent intent = (Intent) XposedHelpers.getObjectField(methodHookParam.args[2], "intent");
                    if(isFCMIntent(intent)){
                        String target = intent.getComponent() == null ? intent.getPackage() : intent.getComponent().getPackageName();
                        if(targetIsAllow(target)){
                            XposedHelpers.callStaticMethod(BroadcastQueueInjector,"checkAbnormalBroadcastInQueueLocked", methodHookParam.args[1], methodHookParam.args[0]);
                            printLog("Allow Auto Start: " + target, true);
                            methodHookParam.setResult(true);
                        }
                    }
                }
            });
        }catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e){
            printLog("No Such Method com.android.server.am.BroadcastQueueInjector.checkApplicationAutoStart");
        }
        try{
            // MIUI 13: Hook BroadcastQueueImpl.checkApplicationAutoStart
            Class<?> BroadcastQueueImpl = XposedHelpers.findClass("com.android.server.am.BroadcastQueueImpl",classLoader);
            XposedUtils.findAndHookMethodAnyParam(BroadcastQueueImpl,"checkApplicationAutoStart",new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    Intent intent = (Intent) XposedHelpers.getObjectField(methodHookParam.args[1], "intent");
                    if(isFCMIntent(intent)){
                        String target = intent.getComponent() == null ? intent.getPackage() : intent.getComponent().getPackageName();
                        if(targetIsAllow(target)){
                            XposedHelpers.callMethod(methodHookParam.thisObject, "checkAbnormalBroadcastInQueueLocked", methodHookParam.args[0]);
                            printLog("Allow Auto Start: " + target, true);
                            methodHookParam.setResult(true);
                        }
                    }
                }
            });
        }catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e){
            printLog("No Such Method com.android.server.am.BroadcastQueueImpl.checkApplicationAutoStart");
        }

        try{
            // HyperOS: Hook BroadcastQueueModernStubImpl 的两个方法
            Class<?> BroadcastQueueImpl = XposedHelpers.findClass("com.android.server.am.BroadcastQueueModernStubImpl",classLoader);
            printLog("[fcmfix] start hook com.android.server.am.BroadcastQueueModernStubImpl.checkApplicationAutoStart");
            XposedUtils.findAndHookMethodAnyParam(BroadcastQueueImpl,"checkApplicationAutoStart", new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    Intent intent = (Intent) XposedHelpers.getObjectField(methodHookParam.args[1], "intent");
                    String target = intent.getComponent() == null ? intent.getPackage() : intent.getComponent().getPackageName();
                    if (targetIsAllow(target)) {
                        // 无日志，先放了
                        printLog("[" + intent.getAction() + "]checkApplicationAutoStart package_name: " + target, true);
                        methodHookParam.setResult(true);
//                        if(isFCMIntent(intent)){
//                            printLog("checkApplicationAutoStart package_name: " + target, true);
//                            methodHookParam.setResult(true);
//                        }else{
//                            printLog("[skip][" + intent.getAction() + "]checkApplicationAutoStart package_name: " + target, true);
//                        }

                    }
                }
            });

            printLog("[fcmfix] start hook com.android.server.am.BroadcastQueueModernStubImpl.checkReceiverIfRestricted");
            XposedUtils.findAndHookMethodAnyParam(BroadcastQueueImpl,"checkReceiverIfRestricted", new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    Intent intent = (Intent) XposedHelpers.getObjectField(methodHookParam.args[1], "intent");
                    String target = intent.getComponent() == null ? intent.getPackage() : intent.getComponent().getPackageName();
                    if(targetIsAllow(target)){
                        if(isFCMIntent(intent)){
                            printLog("BroadcastQueueModernStubImpl.checkReceiverIfRestricted package_name: " + target, true);
                            methodHookParam.setResult(false);
                        }
                    }
                }
            });
        }catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e){
            printLog("No Such class com.android.server.am.BroadcastQueueModernStubImpl");
        }

        try {
            // Hook AutoStartManagerServiceStubImpl.isAllowStartService
            Class<?> AutoStartManagerServiceStubImpl = XposedHelpers.findClass("com.android.server.am.AutoStartManagerServiceStubImpl", classLoader);
            XC_MethodHook methodHook = new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    Intent intent = (Intent) methodHookParam.args[1];
                    String target = intent.getComponent().getPackageName();
                    if(targetIsAllow(target)) {
                        // 拿不到action，先放了
                        printLog("[" + intent.getAction() + "]AutoStartManagerServiceStubImpl.isAllowStartService package_name: " + target, true);
                        methodHookParam.setResult(true);
//                        if(isFCMIntent(intent)){
//                            printLog("AutoStartManagerServiceStubImpl.isAllowStartService package_name: " + target, true);
//                            methodHookParam.setResult(true);
//                        }else{
//                            printLog("[skip][" + intent.getAction() + "]AutoStartManagerServiceStubImpl.isAllowStartService package_name: " + target, true);
//                        }
                    }
                }
            };

            printLog("[fcmfix] start hook com.android.server.am.AutoStartManagerServiceStubImpl.isAllowStartService");
            XC_MethodHook.Unhook unhook1 = XposedUtils.tryFindAndHookMethod(AutoStartManagerServiceStubImpl, "isAllowStartService", 3, methodHook);
            XC_MethodHook.Unhook unhook2 = XposedUtils.tryFindAndHookMethod(AutoStartManagerServiceStubImpl, "isAllowStartService", 4, methodHook);
            if(unhook1 == null && unhook2 == null){
                throw new NoSuchMethodError();
            }
        } catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e){
            printLog("No Such Class com.android.server.am.AutoStartManagerServiceStubImpl.isAllowStartService");
        }

        try {
            // Hook SmartPowerService.shouldInterceptBroadcast
            Class<?> SmartPowerService = XposedHelpers.findClass("com.android.server.am.SmartPowerService", classLoader);

            printLog("[fcmfix] start hook com.android.server.am.SmartPowerService.shouldInterceptBroadcast");
            XposedUtils.findAndHookMethodAnyParam(SmartPowerService, "shouldInterceptBroadcast", new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    Intent intent = (Intent) XposedHelpers.getObjectField(methodHookParam.args[1], "intent");
                    String target = intent.getComponent() == null ? intent.getPackage() : intent.getComponent().getPackageName();
                    if(targetIsAllow(target)) {
                        if(isFCMIntent(intent)){
                            printLog("SmartPowerService.shouldInterceptBroadcast package_name: " + target, true);
                            methodHookParam.setResult(false);
                        }
                    }
                }
            });
        } catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e){
            printLog("No Such Class com.android.server.am.SmartPowerService");
        }

        try{
            // OOS 15 / ColorOS 15: Hook OplusAppStartupManager.shouldPreventSendReceiverReal
            Method method = XposedUtils.findMethod(XposedHelpers.findClass("com.android.server.am.OplusAppStartupManager",classLoader),"shouldPreventSendReceiverReal",4);
            XposedBridge.hookMethod(method,new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam methodHookParam) {
                    if(methodHookParam.args[0] != null && XposedHelpers.getObjectField(methodHookParam.args[0],"intent") != null){
                        Intent intent = (Intent)XposedHelpers.getObjectField(methodHookParam.args[0],"intent");
                        if(isFCMIntent(intent) && targetIsAllow(intent.getPackage())){
                            methodHookParam.setResult(false);
                        }
                    }
                }
            });
        } catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e) {
            printLog("No Such Method com.android.server.am.OplusAppStartupManager.shouldPreventSendReceiverReal");
        }
    }

    /**
     * Hook MIUI 的电源策略管理器，移除对 FCM 服务的拦截
     * 针对 MIUI 13 的 SmartPowerPolicyManager
     */
    protected void startHookRemovePowerPolicy(){
        try {
            // MIUI 13: Hook SmartPowerPolicyManager.shouldInterceptService
            Class<?> AutoStartManagerService = XposedHelpers.findClass("com.miui.server.smartpower.SmartPowerPolicyManager",classLoader);
            XposedUtils.findAndHookMethodAnyParam(AutoStartManagerService,"shouldInterceptService",new XC_MethodHook() {

                @Override
                protected void afterHookedMethod(MethodHookParam param) {
                    Intent intent = (Intent) param.args[0];
                    if("com.google.firebase.MESSAGING_EVENT".equals(intent.getAction())){
                        String target = intent.getComponent() == null ? intent.getPackage() : intent.getComponent().getPackageName();
                        if(targetIsAllow(target)){
                            printLog("Disable MIUI Intercept: " + target, true);
                            param.setResult(false);
                        }
                    }
                }
            });
        } catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e) {
            printLog("No Such Method com.miui.server.smartpower.SmartPowerPolicyManager.shouldInterceptService");
        }
    }
}

File: app/src/main/kotlin/sumicya/fcmself/mods/AutoStartFix.kt
package sumicya.fcmself.mods

import android.content.Intent

import io.github.libxposed.api.XposedInterface

import sumicya.fcmself.FcmselfModule
import sumicya.fcmself.core.FcmselfLog
import sumicya.fcmself.core.Push
import sumicya.fcmself.hook.Hooks
import sumicya.fcmself.hook.Reflect

/**
 * AutoStartFix —— 自启动限制修复。
 *
 * 放行被 ROM 拦下的推送广播，允许已停止的应用被 FCM 唤醒。覆盖：
 * - MIUI 12：`BroadcastQueueInjector.checkApplicationAutoStart`
 * - MIUI 13：`BroadcastQueueImpl.checkApplicationAutoStart`
 * - HyperOS：`BroadcastQueueModernStubImpl.checkApplicationAutoStart` / `checkReceiverIfRestricted`
 * - MIUI / HyperOS：`AutoStartManagerServiceStubImpl.isAllowStartService`、
 *   `SmartPowerService.shouldInterceptBroadcast`、`SmartPowerPolicyManager.shouldInterceptService`
 * - ColorOS / OxygenOS：`OplusAppStartupManager.shouldPreventSendReceiverReal`
 *
 * 统一介入条件（自由化基线）：被拦截对象必须是「目标明确的推送广播」。
 * 定位 Intent 不再依赖固定下标，而是用 [PushArgs] 按两种载体形态扫描实参。
 * 行为变化（相对旧版）：HyperOS 的 checkApplicationAutoStart 与 isAllowStartService
 * 原本对**任意**定向广播放行，本版收敛到推送族；SmartPowerPolicyManager 原本只认
 * MESSAGING_EVENT，本版放宽到完整推送族。
 */
internal class AutoStartFix(api: XposedInterface, classLoader: ClassLoader) :
    FcmselfModule(api, classLoader, "AutoStartFix") {

    init {
        point("MIUI12 BroadcastQueueInjector") {
            hookQueueAutoStart("com.android.server.am.BroadcastQueueInjector", instanceCall = false)
        }
        point("MIUI13 BroadcastQueueImpl") {
            hookQueueAutoStart("com.android.server.am.BroadcastQueueImpl", instanceCall = true)
        }
        point("HyperOS BroadcastQueueModernStubImpl") { hookHyperos() }
        point("AutoStartManagerServiceStubImpl#isAllowStartService") { hookIsAllowStartService() }
        point("SmartPowerService#shouldInterceptBroadcast") { hookSmartPower() }
        point("OplusAppStartupManager#shouldPreventSendReceiverReal") { hookOplus() }
        point("SmartPowerPolicyManager#shouldInterceptService") { hookMiuiPowerPolicy() }
    }

    /**
     * MIUI 12/13 的 checkApplicationAutoStart：推送广播放行（返回 true），
     * 并按旧版行为补调 ROM 的「异常广播检查」留下等价的安全记录。
     * 两者仅在被拦截记录的取参与检查调用的形态上不同（实例/静态）。
     */
    private fun hookQueueAutoStart(className: String, instanceCall: Boolean) {
        val clazz = Reflect.findClass(className, classLoader)
        val method = Reflect.findMethodMostParams(clazz, "checkApplicationAutoStart")
            ?: throw NoSuchMethodError("$className#checkApplicationAutoStart")
        Hooks.hook(api, method) { chain ->
            val args = chain.args.toTypedArray()
            val intent = PushArgs.find(args) ?: return@hook chain.proceed()
            if (instanceCall) {
                Reflect.callMethod(chain.thisObject!!, "checkAbnormalBroadcastInQueueLocked", args.getOrNull(0))
            } else {
                Reflect.callStaticMethod(clazz, "checkAbnormalBroadcastInQueueLocked", args.getOrNull(1), args.getOrNull(0))
            }
            FcmselfLog.log("Allow Auto Start: ${Push.targetOf(intent)}", true)
            true
        }
    }

    /** HyperOS：自启动检查恒放行（true）、接收器限制检查恒放行（false），仅对推送广播生效。 */
    private fun hookHyperos() {
        val className = "com.android.server.am.BroadcastQueueModernStubImpl"
        val clazz = Reflect.findClass(className, classLoader)
        val autoStart = Reflect.findMethodMostParams(clazz, "checkApplicationAutoStart")
        val receiverRestricted = Reflect.findMethodMostParams(clazz, "checkReceiverIfRestricted")

        autoStart?.let { method ->
            Hooks.hook(api, method) { chain ->
                val intent = PushArgs.find(chain.args.toTypedArray()) ?: return@hook chain.proceed()
                FcmselfLog.log("checkApplicationAutoStart package_name: ${Push.targetOf(intent)}", true)
                true
            }
        }
        receiverRestricted?.let { method ->
            Hooks.hook(api, method) { chain ->
                val intent = PushArgs.find(chain.args.toTypedArray()) ?: return@hook chain.proceed()
                FcmselfLog.log("checkReceiverIfRestricted package_name: ${Push.targetOf(intent)}", true)
                false
            }
        }
        if (autoStart == null && receiverRestricted == null) {
            throw NoSuchMethodError("$className#checkApplicationAutoStart / checkReceiverIfRestricted")
        }
    }

    /** MIUI / HyperOS：isAllowStartService（3 参或 4 参）——推送 Intent 即放行。 */
    private fun hookIsAllowStartService() {
        val className = "com.android.server.am.AutoStartManagerServiceStubImpl"
        val clazz = Reflect.findClass(className, classLoader)
        val method = Reflect.findMethodByParamCount(clazz, "isAllowStartService", 4)
            ?: Reflect.findMethodByParamCount(clazz, "isAllowStartService", 3)
            ?: throw NoSuchMethodError("$className#isAllowStartService")
        Hooks.hook(api, method) { chain ->
            val intent = chain.args.filterIsInstance<Intent>().firstOrNull { Push.isTargetedPush(it) }
            if (intent != null) {
                FcmselfLog.log("isAllowStartService package_name: ${Push.targetOf(intent)}", true)
                return@hook true
            }
            chain.proceed()
        }
    }

    /** MIUI / HyperOS：SmartPowerService.shouldInterceptBroadcast —— 推送广播不拦截。 */
    private fun hookSmartPower() {
        val className = "com.android.server.am.SmartPowerService"
        val clazz = Reflect.findClass(className, classLoader)
        val method = Reflect.findMethodMostParams(clazz, "shouldInterceptBroadcast")
            ?: throw NoSuchMethodError("$className#shouldInterceptBroadcast")
        Hooks.hook(api, method) { chain ->
            val intent = PushArgs.find(chain.args.toTypedArray()) ?: return@hook chain.proceed()
            FcmselfLog.log("SmartPowerService.shouldInterceptBroadcast package_name: ${Push.targetOf(intent)}", true)
            false
        }
    }

    /** ColorOS / OxygenOS：shouldPreventSendReceiverReal —— 推送广播放行（false = 不阻止）。 */
    private fun hookOplus() {
        val className = "com.android.server.am.OplusAppStartupManager"
        val clazz = Reflect.findClass(className, classLoader)
        val method = Reflect.findMethodByParamCount(clazz, "shouldPreventSendReceiverReal", 4)
            ?: throw NoSuchMethodError("$className#shouldPreventSendReceiverReal")
        Hooks.hook(api, method) { chain ->
            if (PushArgs.find(chain.args.toTypedArray()) != null) false else chain.proceed()
        }
    }

    /** MIUI 13：SmartPowerPolicyManager.shouldInterceptService —— 推送不拦截（after 改写结果）。 */
    private fun hookMiuiPowerPolicy() {
        val className = "com.miui.server.smartpower.SmartPowerPolicyManager"
        val clazz = Reflect.findClass(className, classLoader)
        val method = Reflect.findMethodMostParams(clazz, "shouldInterceptService")
            ?: throw NoSuchMethodError("$className#shouldInterceptService")
        Hooks.hook(api, method) { chain ->
            val result = chain.proceed()
            val intent = chain.args.filterIsInstance<Intent>().firstOrNull { Push.isTargetedPush(it) }
            if (intent != null) {
                FcmselfLog.log("Disable MIUI Intercept: ${Push.targetOf(intent)}", true)
                false
            } else {
                result
            }
        }
    }
}
